ZHA network settings and backup/restore interface

puddly opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->
Add a new tab to the ZHA interface to show the current network settings to allow them to be backed up/restored.

Unmerged Core PR: https://github.com/home-assistant/core/pull/75791

## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

Testing this requires ZHA to be set up with a Zigbee radio. I've included screenshots below of the various interactions (dialogs moved around with dev tools to show more elements at once):

<details>
<summary><strong>Expand screenshots</strong></summary>
<img width="735" alt="image" src="https://user-images.githubusercontent.com/32534428/181386101-0b66f7f5-a51a-49e6-9baa-1997ff648dbb.png">

<img width="561" alt="image" src="https://user-images.githubusercontent.com/32534428/181386433-34e4d002-2d7c-47e7-a255-fe38989cfa09.png">

<img width="420" alt="image" src="https://user-images.githubusercontent.com/32534428/181386493-833dfbf8-22b8-43b4-aaa2-205f3b20e068.png">

</details>

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

puddly committed:
WIP

puddly committed:
Fully implement backup + restore

puddly committed:
Use the `Record` type for ZHA network setting objects

puddly committed:
Try to get the dialog to close

puddly committed:
Fix dialog not opening after being closed once

puddly committed:
Sort backups by the current time

puddly committed:
Display a message when no backups exist

puddly committed:
Handle incomplete backups

puddly committed:
Show an error dialog when a restore fails

puddly committed:
Save a reference to the backup file to populate the filename

puddly committed:
Reformat

puddly committed:
Use `ha-progress-button` instead of `ha-circular-progress`

puddly committed:
Space out the backup and restore buttons

puddly committed:
Allow overriding the coordinator IEEE address for EZSP coordinators

puddly committed:
Remove leftover debugging logic

puddly committed:
Rename `data` to `backup` in the restore command

puddly committed:
Do not modify the backup JSON within the frontend

puddly committed:
Clean up UI and add translations for some elements

puddly committed:
Run lint

puddly committed:
Run prettier

frenck commented:
Backend has been merged.

zsarnett commented on src/panels/config/integrations/integration-panels/zha/dialog-zha-restore-backup.ts:
```diff
@@ -0,0 +1,368 @@
+import { haStyle, haStyleDialog } from "../../../../../resources/styles";
+import { HomeAssistant } from "../../../../../types";
+
+import { HaRadio } from "../../../../../components/ha-radio";
+import { HaCheckbox } from "../../../../../components/ha-checkbox";
```
```suggestion
import type { HaRadio } from "../../../../../components/ha-radio";
import type { HaCheckbox } from "../../../../../components/ha-checkbox";
import from "../../../../../components/ha-radio";
import from "../../../../../components/ha-checkbox";
```

This makes sure to also import the element itself

zsarnett commented on src/panels/config/integrations/integration-panels/zha/dialog-zha-restore-backup.ts:
```diff
@@ -0,0 +1,368 @@
+import {
```
Can you remove all the empty lines in the import section

zsarnett commented on src/panels/config/integrations/integration-panels/zha/dialog-zha-restore-backup.ts:
```diff
@@ -0,0 +1,368 @@
+    fireEvent(this, "dialog-closed", { dialog: this.localName });
+  }
+
+  private _formatBackupLabel(backup: ZHANetworkBackup): string {
```
`private` methods need to be moved below any `public` or `protected`

zsarnett commented on src/panels/config/integrations/integration-panels/zha/dialog-zha-restore-backup.ts:
```diff
@@ -0,0 +1,368 @@
+    }, ${backup.network_info.source})`;
+  }
+
+  private _shouldOverwriteCoordinatorIEEEAddress(): boolean {
```
same

zsarnett commented on src/panels/config/integrations/integration-panels/zha/dialog-zha-restore-backup.ts:
```diff
@@ -0,0 +1,368 @@
+    return true;
+  }
+
+  private _canOverwriteCoordinatorIEEEAddress(): boolean {
```
same

zsarnett commented on src/panels/config/integrations/integration-panels/zha/dialog-zha-restore-backup.ts:
```diff
@@ -0,0 +1,368 @@
+                    ></ha-file-upload>
+                  </div>
+                `
+              : nothing
```
```suggestion
              : ""
```

zsarnett commented on src/panels/config/integrations/integration-panels/zha/dialog-zha-restore-backup.ts:
```diff
@@ -0,0 +1,368 @@
+                      )}
+                    </div>
+                  `
+              : nothing
```
```suggestion
              : ""
```

zsarnett commented on src/panels/config/integrations/integration-panels/zha/dialog-zha-restore-backup.ts:
```diff
@@ -0,0 +1,368 @@
+                    </ha-formfield>
+                  </div>
+                `
+              : nothing
```
```suggestion
              : ""
```

zsarnett commented on src/panels/config/integrations/integration-panels/zha/zha-network-page.ts:
```diff
@@ -0,0 +1,203 @@
+  ZHANetworkSettings,
+  ZHANetworkBackupAndMetadata,
+} from "../../../../../data/zha";
+
```
```suggestion
```

puddly committed:
Clean up style and member function organization

puddly committed:
Run prettier

puddly commented:
I believe I've addressed all of the issues you mentioned.

Is there any `pre-commit`-like hook for the frontend that auto-formats everything properly, like there is for HA Core?

zsarnett commented:
Ill take a look this week to try and test it. First review is usually just a quick glance.

Hedda commented:
Nice! Please remember that ZHA docs need update for this
